### ISSUE TITLE
rd query parameter is not url-encoded

### ISSUE THREAD

**Issue by Drakulix** [authelia/authelia#476]
First off: Thanks for this awesome project, it works nicely in my traefik setup.

I have an issue when redirecting from more complicated url's then my subdomains. (Redirecting on specific resources instead of the whole page.)
When the redirection url contains a query string, it terminates the `rd` variable, because it is not escaped, which then results in a redirect to the original site without it's query variables.

E.g. in my setup I have something like `https://redacted.org/auth/authorize?response_type=code&...`, which gets redirected to authelia, which results in `https://login.redacted.org/#/?rd=https://redacted.org/auth/authorize?response_type=code&...`.
After authentication I am redirected to `https://redacted.org/auth/authorize` missing any previous query arguments.

I suppose the redirection url should be escaped to be url-safe once extracted from traefiks headers and decoded again for the actual redirect.

**Comment by clems4ever** [comment]
Interesting case @Drakulix, I will see if I can reproduce it and let you know. Thanks in the meantime.

**Comment by tlvenn** [comment]
I am facing the same issue and as pointed in #500, it's not even just with a complicated redirect url / querystring, a simple domain with Traefik 2.0 will not work as `https://` will be replaced with `https:/`

**Comment by 726a67** [comment]
Can confirm the same behavior experienced by @tlvenn using Authelia 4.x in conjunction with Traefik 2.x.

**Comment by nightah** [comment]
The current integration tests are with Traefik 1.x, I'll see if I can have a look at Traefik 2.0 and create a suite for this too. There very likely could be some additional configuration that needs to occur in Traefik or code changes in Authelia to support this.

I would suggest for now sticking to Traefik 1.x, nginx or HAproxy until we have had the opportunity to review this and provide feedback.

**Comment by tlvenn** [comment]
Hi @nightah,

I think it safe to assume that if the query parameter is properly encoded, it will actually solve the compat issue with Traefik 2.0. I am not familiar with golang but I assume an http lib would provide the function utility to do the encoding / decoding.

It should be fairly quick therefore to fix this issue, no saying that having a proper test suite for Traefik 2.0 is not valuable but we should be pragmatic and it makes perfect sense that the query param needs to be encoded in any case.

**Comment by nightah** [comment]
Hi @tlvenn while you're probably not wrong, the point I didn't explicitly state is that I do not have a Traefik 2.x environment to be able to test/confirm any fix.

I actually have very little experience with Traefik at all, so I would need to look at changes from 1.x to 2.x to ensure that we can setup a reproducible environment that will serve us to test any pending and future changes.

If you don't mind providing your configuration/an example for Traefik 2.x, including both the Authelia portal and a protected endpoint it will definitely speed up the process for us to be able to provide support and assistance.

With this I can quickly and easily setup both the suite for the integration tests and look to address what is suspected to be the url encoding issue.

**Comment by tlvenn** [comment]
What i am trying to say is that we should not confuse the issue at hand with providing Traefik 2.0 integration tests.

The query parameter should be properly encoded no matter what upstream proxy is used and that should be the sole focus of this issue.

We should open a new issue to add integration tests for Traefik 2.0 separately.

**Comment by nightah** [comment]
@tlvenn I haven't been able to replicate this issue with Traefik 1.7.20 as per #500.

Could you please provide your configuration so I can attempt to replicate the issue please?

**Comment by clems4ever** [comment]
@tlvenn , I will provide the fix of URL encoding this week. @nightah , I think it should solve most issues and then, when it's implemented, we can see if there is still an issue with Traefik 2.0.

**Comment by nightah** [comment]
@tlvenn/@clems4ever looks like the issue is limited to Traefik2 when utilising the docker provider and specifically when defining the authelia middleware with labels.

@clems4ever initial implementation for Traefik 2.x was with a file based provider because I was unsure if there would be any conflict with Traefik 1.x labels, commit can be viewed at https://github.com/authelia/authelia/commit/55612957aba6a71244eb0e1a8a3842137d317b35. This resulted in https://buildkite.com/authelia/authelia/builds/126#c1a31cd6-556e-46c7-905a-3ca9af38b1ba.

After this I refactored from a file based provider to utilise the docker provider along with the respective Traefik 2.x labels commit can be viewed at https://github.com/authelia/authelia/commit/86fdf4dc4949856dbb5adcff8a43a15a21e753e0. This resulted in https://buildkite.com/authelia/authelia/builds/127#5aca33da-2f56-4111-81af-873b185e3969.

Looking at the Traefik logs, it appears the Referer and RequestURI is being stripped when utilising the docker provider.

With further investigation, it looks like it's actually **ONLY** the middleware definition via the docker provider that causes this specific symptom.

So if we move the middleware definition to be a file based provider and reference that from the container with a label like such: https://github.com/authelia/authelia/commit/34454af65e4fd7301604b916a200c9cee8d60b21 we have Authelia working with Traefik 2.x and are able to utilise all of the fancy service discovery, etc associated with that provider, you can see the test results at https://buildkite.com/authelia/authelia/builds/128#0e1dee49-56d5-44bb-bdfc-42443ac13654.

Clearly there's something odd going on with Traefik because the behaviour differs depending on how the routes are defined (read: providers). I'd be curious to see if URL encoding actually resolves this issue with the middlewares, but I'll leave that to you @clems4ever, please utilise the branch as you see fit.

For anybody that wants to utilise Traefik 2.x right now, the configuration discussed and detailed above work.

**Comment by 726a67** [comment]
I'm also interested to see if the URL encoding addresses the problem when utilizing labels to define the authelia middleware, as I was hoping to have minimal dependency on configuration files for Traefik 2.x (my current config is entirely label-based).

**Comment by nightah** [comment]
For what it's worth, I did test URL encoding the label which is what some users had done in the past when we were initially getting the Traefik suite working. Unfortunately that did not work.

**Comment by tlvenn** [comment]
For info, my setup with Traefik 2.0 is with K8s with both providers.

**Comment by nightah** [comment]
Thanks for the info.

@clems4ever after performing some more testing I've normalised all the Traefik configuration across both the 1.x and 2.x suites to utilise labels in all cases except for the middleware in Traefik 2.x.

Given the above change it's safe to say there was a some sort of regression in Traefik 1.x -> 2.x, however, this might be addressed with URL encoding.

**Comment by clems4ever** [comment]
Authelia (in master) now url encodes the URL passed to rd so this issue should be solved.
@Drakulix , @tlvenn , can you please try and let us know.

**Comment by nightah** [comment]
I've just performed the test again with Traefik2 along with the URL encoding changes and was unsuccessful.

@726a67 if you can test too when you get an opportunity it'd be great to have your feedback too.

If the responses are in the negative, I would suggest lodging an issue upstream with Traefik as it seems quite limited in terms of the symptoms and reproduction.
Using a file based middleware still works without issues, Based on my testing I believe this is limited to middlewares defined with docker labels.

**Comment by 726a67** [comment]
@nightah at least for me, the issue appears to be resolved. Again, my Traefik 2.x config is exclusively label-based, including an Authelia 4.x forwardAuth middleware. Prior to this latest change, certain clients (specifically mobile) would would experience the issue described by @tlvenn:

> 
> 
> I am facing the same issue and as pointed in #500, it's not even just with a complicated redirect url / querystring, a simple domain with Traefik 2.0 will not work as `https://` will be replaced with `https:/`

After pulling the latest Authelia image, my clients behave as expected, with the proper redirect URL now being passed/utilized.

**Comment by nightah** [comment]
Do you mind sharing your configuration please for the authelia portal and the protected endpoint?

I'm just wondering if there's perhaps something I'm missing here, I would be much happier if this is just misconfiguration on my side.

**Comment by 726a67** [comment]
Sure; here's the relevant snippet:

`version: '3'

networks:
  traefik:
    external: true

traefik2:
    image: traefik
    container_name: traefik2
    environment:
      - CF_API_EMAIL=${CF_API_EMAIL}
      - CF_API_KEY=${CF_API_KEY}
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
      - ${CONFIG_DIR}/traefik2/config/acme.json:/acme.json
    networks:
      - traefik
    ports:
      - "80:80"
      - "443:443"
    command:
      - '--api.dashboard=true'
      - '--log.level=DEBUG'
      - '--entrypoints.http.address=:80'
      - '--entrypoints.https.address=:443'
      - '--providers.docker'
      - '--certificatesresolvers.cloudflare.acme.email=${CF_API_EMAIL}'
      - '--certificatesresolvers.cloudflare.acme.storage=/acme.json'
      - '--certificatesresolvers.cloudflare.acme.dnschallenge=true'
      - '--certificatesresolvers.cloudflare.acme.dnschallenge.provider=cloudflare'
      - '--certificatesResolvers.cloudflare.acme.dnsChallenge.resolvers=1.1.1.1:53,1.0.0.1:53'
    labels:
      - 'traefik.http.routers.http-catchall.rule=hostregexp(`{host:.+}`)'
      - 'traefik.http.routers.http-catchall.entrypoints=http'
      - 'traefik.http.routers.http-catchall.middlewares=redirect-to-https'
      - 'traefik.http.middlewares.redirect-to-https.redirectscheme.scheme=https'
      - 'traefik.http.routers.letsencrypt.tls.certresolver=cloudflare'
      - 'traefik.http.routers.letsencrypt.tls.domains[0].main=${DOMAIN}'
      - 'traefik.http.routers.letsencrypt.tls.domains[0].sans=*.${DOMAIN}'
      - 'traefik.http.middlewares.authelia.forwardauth.address=http://authelia:9091/api/verify?rd=https://auth.${DOMAIN}'
      - 'traefik.http.middlewares.authelia.forwardauth.tls.insecureSkipVerify=true'
      # Dashboard
      - 'traefik.http.routers.traefik.rule=Host(`traefik2.${DOMAIN}`)'
      - 'traefik.http.routers.traefik.tls=true'
      - 'traefik.http.routers.traefik.entrypoints=https'
      - 'traefik.http.routers.traefik.service=api@internal'
      - 'traefik.http.routers.traefik.middlewares=authelia@docker'
    restart: unless-stopped

authelia:
    image: authelia/authelia:master
    container_name: authelia
    environment:
      - NODE_TLS_REJECT_UNAUTHORIZED=1
    volumes:
      - ${CONFIG_DIR}/authelia/config:/etc/authelia
    networks:
      - traefik
    labels:
      - 'traefik.http.routers.authelia.rule=Host(`auth.${DOMAIN}`)'
      - 'traefik.http.routers.authelia.tls=true'
      - 'traefik.http.routers.authelia.entrypoints=https'
      - 'traefik.http.routers.authelia.middlewares=authelia@docker'
    depends_on:
      - traefik2
    restart: unless-stopped

filebrowser:
    image: filebrowser/filebrowser
    container_name: filebrowser
    environment:
      - TZ=${TZ}
      - PUID=${PUID}
      - PGID=${PGID}
    volumes:
      - ${CONFIG_DIR}/filebrowser/config/filebrowser.json:/.filebrowser.json
      - ${CONFIG_DIR}/filebrowser/config/filebrowser.db:/filebrowser.db
    networks:
      - traefik
    labels:
      - 'traefik.http.routers.filebrowser.rule=Host(`filebrowser.${DOMAIN}`)'
      - 'traefik.http.routers.filebrowser.tls=true'
      - 'traefik.http.routers.filebrowser.entrypoints=https'
      - 'traefik.http.routers.filebrowser.middlewares=authelia@docker'
    depends_on:
      - traefik2
    restart: unless-stopped`

**Comment by 726a67** [comment]
Sure; here is the relevant snippet:

```
version: '3'

networks:
  traefik:
    external: true

traefik2:
    image: traefik
    container_name: traefik2
    environment:
      - CF_API_EMAIL=${CF_API_EMAIL}
      - CF_API_KEY=${CF_API_KEY}
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
      - ${CONFIG_DIR}/traefik2/config/acme.json:/acme.json
    networks:
      - traefik
    ports:
      - "80:80"
      - "443:443"
    command:
      - '--api.dashboard=true'
      - '--log.level=DEBUG'
      - '--entrypoints.http.address=:80'
      - '--entrypoints.https.address=:443'
      - '--providers.docker'
      - '--certificatesresolvers.cloudflare.acme.email=${CF_API_EMAIL}'
      - '--certificatesresolvers.cloudflare.acme.storage=/acme.json'
      - '--certificatesresolvers.cloudflare.acme.dnschallenge=true'
      - '--certificatesresolvers.cloudflare.acme.dnschallenge.provider=cloudflare'
      - '--certificatesResolvers.cloudflare.acme.dnsChallenge.resolvers=1.1.1.1:53,1.0.0.1:53'
    labels:
      - 'traefik.http.routers.http-catchall.rule=hostregexp(`{host:.+}`)'
      - 'traefik.http.routers.http-catchall.entrypoints=http'
      - 'traefik.http.routers.http-catchall.middlewares=redirect-to-https'
      - 'traefik.http.middlewares.redirect-to-https.redirectscheme.scheme=https'
      - 'traefik.http.routers.letsencrypt.tls.certresolver=cloudflare'
      - 'traefik.http.routers.letsencrypt.tls.domains[0].main=${DOMAIN}'
      - 'traefik.http.routers.letsencrypt.tls.domains[0].sans=*.${DOMAIN}'
      - 'traefik.http.middlewares.authelia.forwardauth.address=http://authelia:9091/api/verify?rd=https://auth.${DOMAIN}'
      - 'traefik.http.middlewares.authelia.forwardauth.tls.insecureSkipVerify=true'
      # Dashboard
      - 'traefik.http.routers.traefik.rule=Host(`traefik2.${DOMAIN}`)'
      - 'traefik.http.routers.traefik.tls=true'
      - 'traefik.http.routers.traefik.entrypoints=https'
      - 'traefik.http.routers.traefik.service=api@internal'
      - 'traefik.http.routers.traefik.middlewares=authelia@docker'
    restart: unless-stopped

authelia:
    image: authelia/authelia:master
    container_name: authelia
    environment:
      - NODE_TLS_REJECT_UNAUTHORIZED=1
    volumes:
      - ${CONFIG_DIR}/authelia/config:/etc/authelia
    networks:
      - traefik
    labels:
      - 'traefik.http.routers.authelia.rule=Host(`auth.${DOMAIN}`)'
      - 'traefik.http.routers.authelia.tls=true'
      - 'traefik.http.routers.authelia.entrypoints=https'
      - 'traefik.http.routers.authelia.middlewares=authelia@docker'
    depends_on:
      - traefik2
    restart: unless-stopped

filebrowser:
    image: filebrowser/filebrowser
    container_name: filebrowser
    environment:
      - TZ=${TZ}
      - PUID=${PUID}
      - PGID=${PGID}
    volumes:
      - ${CONFIG_DIR}/filebrowser/config/filebrowser.json:/.filebrowser.json
      - ${CONFIG_DIR}/filebrowser/config/filebrowser.db:/filebrowser.db
    networks:
      - traefik
    labels:
      - 'traefik.http.routers.filebrowser.rule=Host(`filebrowser.${DOMAIN}`)'
      - 'traefik.http.routers.filebrowser.tls=true'
      - 'traefik.http.routers.filebrowser.entrypoints=https'
      - 'traefik.http.routers.filebrowser.middlewares=authelia@docker'
    depends_on:
      - traefik2
    restart: unless-stopped
```

**Comment by nightah** [comment]
Well it really was a simple misconfiguration issue it appears, the issue I was having was due to single quotes vs double quotes and how that's interpreted by Docker.

Thanks @726a67!

**Comment by clems4ever** [comment]
Closing now, please re-open the issue if some of you are still facing the issue after the bump.

**Comment by tlvenn** [comment]
Finally able to report that the encoding fixed my issue with Traefik 2.1, thanks guys.

**Comment by clems4ever** [comment]
Good to hear @tlvenn .
